src/environmentbase/template.py

```diff
@@ -537,6 +506,76 @@ def add_cloudwatch_alarm(self,
+    def image_map_name(image_label):
+        """
+        Each image map is a concatenation of a custom label and a prefix. This function
+        provides a consistant way to get the correct image map name from the custom label.
+        @returns string
+        """
+        return Template.IMAGE_MAP_PREFIX + image_label.title()
+
+    def get_instancetype_param(self, instance_type, image_name, layer_name):
+        """
+        @param instance_type [string] Default instance_type value.
+        @param image_name [string] Key from config.image_map, used to create Parameter name
+        @param layer_name [string] Functional name for the template, also used in Parameter name
+        Returns a troposphere.Parameter configured to take instance types.  The parameter is only
+        created if it does not already exist.
+        """
+        if not isinstance(instance_type, basestring):
+            raise Exception('Template.get_instancetype_param::instance_type should be string')
+
+        param_name = layer_name.title() + 'InstanceTypeFor' + image_name.title()
+
+        # Return parameter if it already exists
+        if param_name in self.parameters:
+            return self.parameters.get(param_name)
+
+        sorted_instancetypes = sorted(Template.instancetype_to_arch.keys())
+        return self.add_parameter(Parameter(
+            param_name,
+            Description='AWS instance type',
+            Default=instance_type,
+            Type='String',
+            AllowedValues=sorted_instancetypes))
+
+    def get_ami(self, instance_type, image_name, layer_name):
+        """
+        @param instance_type [string] Default instance_type value.
+        @param image_name [string] Key from config.image_map, used to create Parameter name
+        @param layer_name [string] Functional name for the template, also used in Parameter name
+        Returns a CloudFormation expression which will evaluate to an AMI id.  This is done by
+        attaching a single instancetype_to_arch map and a region_to_ami_id map (per image type).
+        Note: Resulting templates can deploy with different instance_types into differnet regions
+        without any modification.
+        """
+        if not isinstance(instance_type, basestring):
+            raise Exception('Template.get_ami::instance_type should be string')
+
+        # Static validation of instance type
+        if instance_type not in Template.instancetype_to_arch.keys():
+            raise Exception('Unrecognized instance type: "%s"' % instance_type)
+
+        # Add input parameter (\w runtime validation) for instance_type
+        instancetype_param = self.get_instancetype_param(instance_type, image_name, layer_name)
+
+        # Lazy-add instancetype_to_arch map
+        if Template.ARCH_MAP not in self.mappings:
+            arch_map = {inst_type: {"Arch": arch} for inst_type, arch in Template.instancetype_to_arch.items()}
+            self.mappings[Template.ARCH_MAP] = arch_map
```

Comment:
I would put this into a different helper function since it has a well defined purpose that is separable from the purpose of this function.